File: scripts/guard-selftest.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Adversarial self-test for the SDD guard.
#
# The guard (check-sdd.rb) is the single enforcement point for the whole
# governance model, so "the guard passes" is only meaningful if the guard is
# also proven to FAIL on tampering. Each case below copies the repository to
# a scratch directory, injects exactly one class of violation and asserts
# that the guard reports the expected error. A mutation the guard fails to
# detect fails this suite.
#
# Run alongside scripts/check-sdd.sh; CI must require both to pass.
# Override the scratch location with ARKDECK_SELFTEST_TMPDIR if needed.

require "fileutils"
require "open3"
require "pathname"
require "tmpdir"

SOURCE_ROOT = Pathname.new(__dir__).parent.expand_path
COPY_EXCLUDES = %w[.git .claude __pycache__].freeze

def copy_repo(destination)
  SOURCE_ROOT.children.each do |entry|
    next if COPY_EXCLUDES.include?(entry.basename.to_s)
    FileUtils.cp_r(entry, destination, preserve: true)
  end
  Dir.glob(File.join(destination, "**", "__pycache__")).each { |dir| FileUtils.rm_rf(dir) }
end

def run_guard(root)
  Open3.capture3("ruby", File.join(root, "scripts", "check-sdd.rb"))
end

def run_relock(root)
  Open3.capture3("ruby", File.join(root, "scripts", "relock-baseline.rb"))
end

Case = Struct.new(:name, :expected_error, :mutation, keyword_init: true)

CASES = [
  Case.new(
    name: "tamper-protected-content",
    expected_error: %r{baseline protected hash mismatch: openspec/constitution\.md},
    mutation: ->(root) { File.write(File.join(root, "openspec/constitution.md"), "\n<!-- tampered -->\n", mode: "a") }
  ),
  Case.new(
    name: "inject-unregistered-protected-file",
    expected_error: %r{baseline omits protected files: .*zzz-injected},
    mutation: ->(root) { File.write(File.join(root, "openspec/contracts/zzz-injected.schema.json"), "{}\n") }
  ),
  Case.new(
    name: "delete-protected-file",
    expected_error: %r{baseline protected path missing: openspec/specs/flashing/spec\.md},
    mutation: ->(root) { FileUtils.rm(File.join(root, "openspec/specs/flashing/spec.md")) }
  ),
  Case.new(
    name: "tamper-file-manifest",
    expected_error: %r{baseline hash mismatch: openspec/baselines/},
    mutation: lambda do |root|
      manifest = Dir.glob(File.join(root, "openspec/baselines/*.files.yaml")).fetch(0)
      text = File.read(manifest)
      # Flip the first hex digit of the first recorded hash.
      File.write(manifest, text.sub(/sha256: ([0-9a-f])/) { "sha256: #{Regexp.last_match(1) == '0' ? '1' : '0'}" })
    end
  ),
  Case.new(
    name: "unsorted-file-manifest",
    expected_error: /baseline file manifest is not path-sorted/,
    mutation: lambda do |root|
      manifest = Dir.glob(File.join(root, "openspec/baselines/*.files.yaml")).fetch(0)
      lines = File.readlines(manifest)
      first = lines.index { |line| line.start_with?("- path: ") }
      # Swap the first two two-line entries so paths are out of order.
      lines[first, 4] = [lines[first + 2], lines[first + 3], lines[first], lines[first + 1]]
      File.write(manifest, lines.join)
    end
  ),
  Case.new(
    name: "drop-acceptance-index-entry",
    expected_error: /acceptance index missing: AC-/,
    mutation: lambda do |root|
      index = File.join(root, "openspec/verification/acceptance-index.txt")
      lines = File.readlines(index)
      victim = lines.index { |line| line.start_with?("AC-") }
      lines.delete_at(victim)
      File.write(index, lines.join)
    end
  ),
  Case.new(
    name: "add-unknown-acceptance-id",
    expected_error: /acceptance index has unknown IDs: AC-ZZZ-999-01/,
    mutation: lambda do |root|
      File.write(File.join(root, "openspec/verification/acceptance-index.txt"), "AC-ZZZ-999-01\n", mode: "a")
    end
  ),
  Case.new(
    name: "duplicate-acceptance-scenario",
    expected_error: /duplicate Acceptance AC-DUMP-001-01/,
    mutation: lambda do |root|
      spec = File.join(root, "openspec/specs/ui-dump/spec.md")
      File.write(spec, "\n#### Scenario: AC-DUMP-001-01 duplicate injection\n\n- GIVEN x\n- WHEN y\n- THEN z\n", mode: "a")
    end
  ),
  Case.new(
    name: "requirement-without-scenario",
    expected_error: /REQ-ZZZ-001 has no Scenario/,
    mutation: lambda do |root|
      spec = File.join(root, "openspec/specs/ui-dump/spec.md")
      File.write(spec, "\n### Requirement: REQ-ZZZ-001 Injected requirement\n\nTHE SYSTEM SHALL be detected.\n", mode: "a")
    end
  ),
  Case.new(
    name: "escalate-task-packet-status",
    expected_error: /ready Task TASK-M0A-001/,
    mutation: lambda do |root|
      packet = File.join(root, "openspec/changes/chg-2026-001-macos-m0a/task-packets/TASK-M0A-001.json")
      File.write(packet, File.read(packet).sub('"status": "draft"', '"status": "ready"'))
    end
  ),
  Case.new(
    name: "tamper-platform-profile",
    expected_error: %r{platform lock hash mismatch: openspec/platforms/linux/profile\.md},
    mutation: ->(root) { File.write(File.join(root, "openspec/platforms/linux/profile.md"), "\n<!-- tampered -->\n", mode: "a") }
  )
].freeze

failures = []
base_tmp = ENV["ARKDECK_SELFTEST_TMPDIR"]
FileUtils.mkdir_p(base_tmp) if base_tmp

Dir.mktmpdir("arkdeck-guard-selftest-", base_tmp) do |tmp|
  # Case 0: the pristine copy must be green, otherwise mutations prove nothing.
  pristine = File.join(tmp, "pristine")
  FileUtils.mkdir_p(pristine)
  copy_repo(pristine)
  stdout, stderr, status = run_guard(pristine)
  if status.success?
    puts "PASS pristine-copy-is-green"
  else
    failures << "pristine-copy-is-green"
    puts "FAIL pristine-copy-is-green — guard must pass on an unmodified tree before mutations mean anything:\n#{stderr}"
  end

  CASES.each do |test_case|
    root = File.join(tmp, test_case.name)
    FileUtils.mkdir_p(root)
    copy_repo(root)
    test_case.mutation.call(root)
    _stdout, stderr, status = run_guard(root)
    if status.success?
      failures << test_case.name
      puts "FAIL #{test_case.name} — guard did not detect the mutation"
    elsif stderr.match?(test_case.expected_error)
      puts "PASS #{test_case.name}"
    else
      failures << test_case.name
      puts "FAIL #{test_case.name} — guard failed, but without the expected error #{test_case.expected_error.inspect}:\n#{stderr}"
    end
  end

  # Round-trip: relock must repair candidate drift and return the guard to
  # green, and must refuse to run against an accepted baseline.
  round_trip = File.join(tmp, "relock-repairs-drift")
  FileUtils.mkdir_p(round_trip)
  copy_repo(round_trip)
  File.write(File.join(round_trip, "openspec/constitution.md"), "\n<!-- candidate drift -->\n", mode: "a")
  _stdout, _stderr, drift_status = run_guard(round_trip)
  relock_out, relock_err, relock_status = run_relock(round_trip)
  _stdout, post_err, post_status = run_guard(round_trip)
  if !drift_status.success? && relock_status.success? && post_status.success?
    puts "PASS relock-repairs-drift"
  else
    failures << "relock-repairs-drift"
    puts "FAIL relock-repairs-drift — drift detected: #{!drift_status.success?}, relock: #{relock_status.success?} (#{relock_err}#{relock_out}), post-relock guard: #{post_status.success?}\n#{post_err}"
  end

  refusal = File.join(tmp, "relock-refuses-accepted-baseline")
  FileUtils.mkdir_p(refusal)
  copy_repo(refusal)
  lock = Dir.glob(File.join(refusal, "openspec/baselines/*.lock.yaml")).fetch(0)
  File.write(lock, File.read(lock).sub(/^status: review$/, "status: accepted"))
  _stdout, refusal_err, refusal_status = run_relock(refusal)
  if !refusal_status.success? && refusal_err.include?("refusing to relock")
    puts "PASS relock-refuses-accepted-baseline"
  else
    failures << "relock-refuses-accepted-baseline"
    puts "FAIL relock-refuses-accepted-baseline — relock must never rewrite an accepted baseline"
  end
end

if failures.empty?
  puts "Guard self-test passed: #{CASES.length + 3} cases."
  exit 0
end

warn "Guard self-test FAILED: #{failures.join(', ')}"
exit 1

File: scripts/guard_selftest.py
#!/usr/bin/env python3.14
"""Adversarial self-test for the ArkDeck SDD guard."""

from __future__ import annotations

import os
import re
import shutil
import subprocess
import sys
import tempfile
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from sdd_protected_set import require_sdd_runtime


require_sdd_runtime()


# The guard (check_sdd.py) is the single enforcement point for the whole
# governance model, so "the guard passes" is only meaningful if the guard is
# also proven to FAIL on tampering. Each case below copies the repository to a
# scratch directory, injects exactly one class of violation and asserts that
# the guard reports the expected error. A mutation the guard fails to detect
# fails this suite.
#
# Run alongside scripts/check-sdd.sh; CI must require both to pass.
# Override the scratch location with ARKDECK_SELFTEST_TMPDIR if needed.
SOURCE_ROOT = Path(__file__).resolve().parent.parent
COPY_EXCLUDES = frozenset((".git", ".claude", "__pycache__"))


def _is_virtualenv(entry: Path) -> bool:
    return entry.is_dir() and (entry / "pyvenv.cfg").is_file()


def copy_repo(destination: Path) -> None:
    for entry in SOURCE_ROOT.iterdir():
        if entry.name in COPY_EXCLUDES or _is_virtualenv(entry):
            continue
        target = destination / entry.name
        if entry.is_dir():
            shutil.copytree(entry, target, copy_function=shutil.copy2)
        else:
            shutil.copy2(entry, target)
    for cache in destination.rglob("__pycache__"):
        if cache.is_dir():
            shutil.rmtree(cache)


def run_python(root: Path, script_name: str) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        [sys.executable, str(root / "scripts" / script_name)],
        check=False,
        capture_output=True,
        text=True,
    )


def run_guard(root: Path) -> subprocess.CompletedProcess[str]:
    return run_python(root, "check_sdd.py")


def run_relock(root: Path) -> subprocess.CompletedProcess[str]:
    return run_python(root, "relock_baseline.py")


@dataclass(frozen=True, slots=True)
class Case:
    name: str
    expected_error: re.Pattern[str]
    mutation: Callable[[Path], None]


def append_text(path: Path, content: str) -> None:
    with path.open("a", encoding="utf-8") as stream:
        stream.write(content)


def replace_text(path: Path, old: str, new: str) -> None:
    path.write_text(path.read_text(encoding="utf-8").replace(old, new, 1), encoding="utf-8")


def manifest_path(root: Path) -> Path:
    return next((root / "openspec/baselines").glob("*.files.yaml"))


def regex_inspect(pattern: re.Pattern[str]) -> str:
    """Render a pattern as /.../ for stable failure output."""

    return f"/{pattern.pattern}/"


def bool_text(value: bool) -> str:
    return "true" if value else "false"


def tamper_file_manifest(root: Path) -> None:
    path = manifest_path(root)
    text = path.read_text(encoding="utf-8")

    def flip(match: re.Match[str]) -> str:
        return f"sha256: {'1' if match.group(1) == '0' else '0'}"

    path.write_text(re.sub(r"sha256: ([0-9a-f])", flip, text, count=1), encoding="utf-8")


def unsort_file_manifest(root: Path) -> None:
    path = manifest_path(root)
    lines = path.read_text(encoding="utf-8").splitlines(keepends=True)
    first = next(index for index, line in enumerate(lines) if line.startswith("- path: "))
    lines[first : first + 4] = lines[first + 2 : first + 4] + lines[first : first + 2]
    path.write_text("".join(lines), encoding="utf-8")


def drop_acceptance_index_entry(root: Path) -> None:
    path = root / "openspec/verification/acceptance-index.txt"
    lines = path.read_text(encoding="utf-8").splitlines(keepends=True)
    victim = next(index for index, line in enumerate(lines) if line.startswith("AC-"))
    del lines[victim]
    path.write_text("".join(lines), encoding="utf-8")


CASES = (
    Case(
        name="tamper-protected-content",
        expected_error=re.compile(r"baseline protected hash mismatch: openspec/constitution\.md"),
        mutation=lambda root: append_text(root / "openspec/constitution.md", "\n<!-- tampered -->\n"),
    ),
    Case(
        name="inject-unregistered-protected-file",
        expected_error=re.compile(r"baseline omits protected files: .*zzz-injected"),
        mutation=lambda root: (root / "openspec/contracts/zzz-injected.schema.json").write_text(
            "{}\n", encoding="utf-8"
        ),
    ),
    Case(
        name="delete-protected-file",
        expected_error=re.compile(r"baseline protected path missing: openspec/specs/flashing/spec\.md"),
        mutation=lambda root: (root / "openspec/specs/flashing/spec.md").unlink(),
    ),
    Case(
        name="tamper-file-manifest",
        expected_error=re.compile(r"baseline hash mismatch: openspec/baselines/"),
        mutation=tamper_file_manifest,
    ),
    Case(
        name="unsorted-file-manifest",
        expected_error=re.compile(r"baseline file manifest is not path-sorted"),
        mutation=unsort_file_manifest,
    ),
    Case(
        name="drop-acceptance-index-entry",
        expected_error=re.compile(r"acceptance index missing: AC-"),
        mutation=drop_acceptance_index_entry,
    ),
    Case(
        name="add-unknown-acceptance-id",
        expected_error=re.compile(r"acceptance index has unknown IDs: AC-ZZZ-999-01"),
        mutation=lambda root: append_text(
            root / "openspec/verification/acceptance-index.txt", "AC-ZZZ-999-01\n"
        ),
    ),
    Case(
        name="duplicate-acceptance-scenario",
        expected_error=re.compile(r"duplicate Acceptance AC-DUMP-001-01"),
        mutation=lambda root: append_text(
            root / "openspec/specs/ui-dump/spec.md",
            "\n#### Scenario: AC-DUMP-001-01 duplicate injection\n\n"
            "- GIVEN x\n- WHEN y\n- THEN z\n",
        ),
    ),
    Case(
        name="requirement-without-scenario",
        expected_error=re.compile(r"REQ-ZZZ-001 has no Scenario"),
        mutation=lambda root: append_text(
            root / "openspec/specs/ui-dump/spec.md",
            "\n### Requirement: REQ-ZZZ-001 Injected requirement\n\n"
            "THE SYSTEM SHALL be detected.\n",
        ),
    ),
    Case(
        name="escalate-task-packet-status",
        expected_error=re.compile(r"ready Task TASK-M0A-001"),
        mutation=lambda root: replace_text(
            root
            / "openspec/changes/chg-2026-001-macos-m0a/task-packets/TASK-M0A-001.json",
            '"status": "draft"',
            '"status": "ready"',
        ),
    ),
    Case(
        name="tamper-platform-profile",
        expected_error=re.compile(r"platform lock hash mismatch: openspec/platforms/linux/profile\.md"),
        mutation=lambda root: append_text(
            root / "openspec/platforms/linux/profile.md", "\n<!-- tampered -->\n"
        ),
    ),
)


def main() -> int:
    failures: list[str] = []
    base_tmp_text = os.environ.get("ARKDECK_SELFTEST_TMPDIR")
    base_tmp = Path(base_tmp_text) if base_tmp_text else None
    if base_tmp is not None:
        base_tmp.mkdir(parents=True, exist_ok=True)

    with tempfile.TemporaryDirectory(prefix="arkdeck-guard-selftest-", dir=base_tmp) as tmp:
        temp_root = Path(tmp)

        # Case 0: the pristine copy must be green, otherwise mutations prove nothing.
        pristine = temp_root / "pristine"
        pristine.mkdir()
        copy_repo(pristine)
        result = run_guard(pristine)
        if result.returncode == 0:
            print("PASS pristine-copy-is-green")
        else:
            failures.append("pristine-copy-is-green")
            print(
                "FAIL pristine-copy-is-green — guard must pass on an unmodified tree "
                f"before mutations mean anything:\n{result.stderr}"
            )

        for test_case in CASES:
            root = temp_root / test_case.name
            root.mkdir()
            copy_repo(root)
            test_case.mutation(root)
            result = run_guard(root)
            if result.returncode == 0:
                failures.append(test_case.name)
                print(f"FAIL {test_case.name} — guard did not detect the mutation")
            elif test_case.expected_error.search(result.stderr):
                print(f"PASS {test_case.name}")
            else:
                failures.append(test_case.name)
                print(
                    f"FAIL {test_case.name} — guard failed, but without the expected error "
                    f"{regex_inspect(test_case.expected_error)}:\n{result.stderr}"
                )

        # Round-trip: relock must repair candidate drift and return the guard to
        # green, and must refuse to run against an accepted baseline.
        round_trip = temp_root / "relock-repairs-drift"
        round_trip.mkdir()
        copy_repo(round_trip)
        append_text(round_trip / "openspec/constitution.md", "\n<!-- candidate drift -->\n")
        drift_result = run_guard(round_trip)
        relock_result = run_relock(round_trip)
        post_result = run_guard(round_trip)
        if (
            drift_result.returncode != 0
            and relock_result.returncode == 0
            and post_result.returncode == 0
        ):
            print("PASS relock-repairs-drift")
        else:
            failures.append("relock-repairs-drift")
            print(
                "FAIL relock-repairs-drift — drift detected: "
                f"{bool_text(drift_result.returncode != 0)}, relock: "
                f"{bool_text(relock_result.returncode == 0)} "
                f"({relock_result.stderr}{relock_result.stdout}), post-relock guard: "
                f"{bool_text(post_result.returncode == 0)}\n{post_result.stderr}"
            )

        refusal = temp_root / "relock-refuses-accepted-baseline"
        refusal.mkdir()
        copy_repo(refusal)
        lock = next((refusal / "openspec/baselines").glob("*.lock.yaml"))
        lock.write_text(
            re.sub(
                r"^status: review$",
                "status: accepted",
                lock.read_text(encoding="utf-8"),
                count=1,
                flags=re.MULTILINE,
            ),
            encoding="utf-8",
        )
        refusal_result = run_relock(refusal)
        if refusal_result.returncode != 0 and "refusing to relock" in refusal_result.stderr:
            print("PASS relock-refuses-accepted-baseline")
        else:
            failures.append("relock-refuses-accepted-baseline")
            print("FAIL relock-refuses-accepted-baseline — relock must never rewrite an accepted baseline")

    if not failures:
        print(f"Guard self-test passed: {len(CASES) + 3} cases.")
        return 0

    print(f"Guard self-test FAILED: {', '.join(failures)}", file=sys.stderr)
    return 1


if __name__ == "__main__":
    raise SystemExit(main())
